Smart(er) resolution of the current working directory.

mitskevich opened the pull request:
First of all, sorry for somewhat big PR, I tried to structure it to be as comprehensible as I can. Let me know if you need any clarifications, I'll gladly provide all needed details.

Issue is as follows:
1. Open Rust project in VSCode.
2. Open any Rust file.
3. Invoke "Go to Symbol in Workspace..." and start typing any symbol name. Notice that the symbol list is working correctly.
4. Close all opened file tabs in VSCode (but not the project folder itself).
5. Repeat step 3. After the symbol name is typed, extension fails with an exception.

This is actually only the tip of the iceberg, there are other different kinds of issues when you:
* Try to invoke Cargo command when there is no file opened in VSCode.
* Try to use symbol list or invoke Cargo command when you have a file opened that doesn't belong to current workspace.
* Try to use symbol list or invoke Cargo command when you are working with the file that lies in the folder that doesn't belong to any Rust project.

Underlying problem is the resolution of current project folder in `PathService.cwd()`. 

Proposed PR attempts to fix it by providing somewhat smarter algorithm. Here is the description from the code commentary:

```
This procedure is complicated by the fact, that the currently opened VSCode workspace may
not be exact Rust project directory, or may even contain several Rust projects.

Let's say we have a workspace opened, and it has a directory structure that looks like this:
- BigSolution/
  - RustProject1/
    - Cargo.toml
    - main.rs
  - RustProject2/
    - Cargo.toml
    - src/
      - someotherfile.sql
      - somesource.rs
  - TODO.txt

If currently active editor is BigSolution/RustProject1/main.rs this function should return
BigSolution/RustProject1/

If currently active editor is BigSolution/RustProject2/src/someotherfile.sql this function
should return BigSolution/RustProject2/

If currently active editor is BigSolution/TODO.txt, or there is no currently active editor
(window has no editors opened at all), then it is ambiguous which project should be the
context of the invoked command. In this case, we return first from the following sequence:

1. Workspace root directory, if it contains Cargo.toml file.

2. Last project directory user has worked with (value that was returned by the last cwd()
    call), if there was one, if it still exists and if it still has Cargo.toml file.

3. First directory in the workspace that has a Cargo.toml file (found by recursive search).

4. Error.
```

I also changed the type of the value returned by `cwd()` from `Promise<string|Error>` to `Promise<string>`. If nothing is found, promise just fails with an `Error`. This approach is more idiomatic and easier to implement elegantly. Because of this, there is a lot of whitespace changes in `src/services/commandService.ts` but that's mostly indentation, the only meaningful thing is the extraction of `if` into a second argument for `then()`.

KalitaAlexey commented:
Hi @mitskevich,

I don't like where we go. It becomes more complex.

I suggest to do some changes.
Activate extension if a workspace (directory) contains **Cargo.toml**.
Look for **Cargo.toml** only in the workspace.

It makes the extension more fast.
It prevents writing of a lot of code.
It allows to use symbol navigation and to execute commands without active editor.
It allows parallel builds (open multiple VSCode windows and here you go).

I like to open my folder with projects and edit files from these projects, so for me it is inconvenient, but there is greater motivation.
If I open each project in a separate window, I would reduce number of opened documents (tabs).
I wouldn't stop current build of some project by starting a build of another project.

What do you think about it?

mitskevich commented:
In my experience, often Rust projects are a small part of some big solution and it is much more convenient to work with whole solution in VSCode than only with the Rust part. It may be a web app written in a completely different language that integrate Rust in some of its modules.

Even if it's a Rust project, it is not always better to have a Cargo.toml file and sources in the root dir. For example, this is what one of my Rust projects contain:
* Deployment/docker configs
* Database setup, fixtures and SQL snippets
* Project management stuff and docs
* Runtime temporary stuff
* Functional tests (test deployment as a whole, not only sources)

So, obviously it's easier to have Rust sources in a subfolder, but still, there is a need at least to quickly switch between files in a whole project.

> It makes the extension more fast.

This PR is at least as fast for the setup you are proposing as current solution is. There is no downside.

> It prevents writing of a lot of code.

I understand that there is a cognitive load to support more code, and it's a perfectly valid point to try keeping things simple. But that does not resolve the actual problem: some of the developers does not dictate how the project they are working with is structured (even if they don't like it). And if this tool can't work with their setup, they simply won't use it.

> It allows to use symbol navigation and to execute commands without active editor.

Same for the proposed PR. Again, if there is a Cargo.toml file in WS root, it's as fast as current solution.

> It allows parallel builds (open multiple VSCode windows and here you go).

Same here. But with your proposed approach it will be _requirement_. With this PR it's just an option.

KalitaAlexey commented:
@mitskevich,

I have an idea. I will describe it later.

KalitaAlexey commented:
@mitskevich,

I want to allow a user to decide what project to build.

# Simple case

If a user opens a directory with Cargo.toml then the directory is the project's directory.
If the user starts a cargo command then the directory will be used as the current working directory for the command.
If the user opens a file, which outside of the current project, the project's directory isn't changed.

# Complex case

Let's assume we have the following directory:
```
Workspace/
    Project 1/
        Cargo.toml
        src/
            main.rs
    Project 2/
        Cargo.toml
        src/
            main.rs
```

If a user opens the directory then the extension isn't activated because there is no **Cargo.toml** in the workspace's directory.

In order to make a directory be the project's directory, the user must open **Cargo.toml** and click on an information message with the text "Use this file to build?" with the choices: "Yes", "Cancel".

# Motivation

It keeps things simple.

It allows a user to work on mixed project.

It allows a user to work on a library and an application in parallel (Use the application's **Cargo.toml** to build, but edit the library's files).

It something is not clear let me now please.

Regards,
Alexey

mitskevich commented:
So, basically, anytime I'll have to switch to file in another project (but in the same solution), I'll have to select Cargo.toml file in this project first and tell VSCode to use it? This prevents me of editing two files from separate projects (in two opened tabs) because linting won't work... Sorry, I don't understand the upsides of this approach compared to PR. I think I'm missing something, please elaborate on your points.

Here is how I understand them:

> It keeps things simple.

It will take similar amount of code to support, but it will also complicate the user experience. User will have to interact with dialogs and do the machine's work. Decision of which project directory is current is not ambiguous 99% of time (just look up Cargo.toml up in the folder structure, computer!). In the remaining 1% it is possible to do the best guess and even if it won't be right, there is nothing to lose.

> It allows a user to work on mixed project.

Again, PR does this too, but with less pressure on user. Most of the times people won't even notice that there may be any restriction on the folder structure.

> It allows a user to work on a library and an application in parallel (Use the application's Cargo.toml to build, but edit the library's files).

Usually bin files are in the subfolder of Rust project, so I don't think there will be any problem with that setup in any of the proposed solutions (even if we don't change anything at all).

---

Again, proposed PR works well for both cases you described. We can add caching to better suit simple case, but I believe that we will need to check existence of Cargo.toml in WS root every time anyway (what if it was moved/deleted?).

KalitaAlexey commented:
@mitskevich, you told some cases, which I hadn't considered.
Let me check it.

KalitaAlexey commented:
I like your solution, but I have a few questions:
Why do we need this?
```
3. First directory in the workspace that has a Cargo.toml file (found by recursive search).
```
I think that the extension should start with:
```
1. Workspace root directory, if it contains Cargo.toml file.
```
When we open a file, **Cargo.toml** is looked up to the file.
When we have no active editors the last **Cargo.toml** is taken.

What do you think about it?

mitskevich commented:
> 3. First directory in the workspace that has a Cargo.toml file (found by recursive search).

When one don't have any file open it is usual to go to a symbol in a workspace (hit Cmd+T, put a first letters of a symbol name, hit Enter). Without this option there will be no way to do that.

Since recursive search is the last resort, it won't affect usual performance. Usually some file _is_ opened, or, as you said, usually there is a Cargo.toml in the workspace root. But in those rare cases, when there isn't (not so rare for me actually), it's better to have that option, then to have nothing at all.

---

As for your second point, it actually works exactly like you described. It looks up Cargo.toml if there is a file opened, otherwise it tries WS root Cargo.toml first thing. If there is none, it tries previous opened project dir. If there is none - recursive search.

KalitaAlexey commented:
You want to look for **Cargo.toml** in nested directories, don't you?

KalitaAlexey commented:
@mitskevich,

Let me know if I am wrong, but you are suggesting the following scenario:
* Open a folder in VSCode.
* Press CTRL+T
* It should find first **Cargo.toml** in nested directories of the folder

mitskevich commented:
@KalitaAlexey, nope, that won't work because extension will not be activated. Usually I find myself trying to find a symbol in the workspace when I'm already working with the project... So, it's a bit different:

* Open a folder in VSCode.
* Work, work, work (with Rust files)...
* Close all file tabs, because I don't need them anymore.
* Now I need to go to my_favourite_function, but I don't remember where it is.
* Hit CTRL+T
* Type in "my_fav"... And now I want VSCode to find it (but instead extension just fails with exception).

KalitaAlexey commented:
@mitskevich Actually extension is still activated.

mitskevich commented:
Extension will be activated in my example (because Rust files were opened prior to the Symbol search). In your example, if there was no Rust file opened, it will be inactive. I'm not proposing fix for that behaviour.

KalitaAlexey commented:
@mitskevich,
You have written:
> Work, work, work (with Rust files)...

If you worked with Rust files the extension would be activated.

KalitaAlexey commented:
@mitskevich,
Let me summarize.
If a workspace's directory contains **Cargo.toml** Rust extension is activated and this **Cargo.toml** is marked as last.
I would like to have the following a project's path resolution:
* If there is an active text editor look for **Cargo.toml** up to workspace's directory.
  * If there is **Cargo.toml** use it.
  * If there is no **Cargo.toml** use the last **Cargo.toml**.
    * If there is no last **Cargo.toml** show an error.
* If there is no active text editor use the last **Cargo.toml**.
  * If there is no last **Cargo.toml** show an error.

Do you agree with me?

mitskevich commented:
Okay, let's go line by line, to make sure we're on the same page...

> If a workspace's directory contains Cargo.toml Rust extension is activated and this Cargo.toml is marked as last.

I don't see any serious reason to activate the extension when there is a Cargo.toml in the root, except for a very rare case when someone opens VSCode and immediately wants to go to the symbol. It's fine if you want it, but my case was not about that.

My case is about much more common example, when you are already working with the project and want to go to the symbol. Right now you can do it _if you have a tab open_. If there are no tabs open, **extension fails**.

> I would like to have the following a project's path resolution:

> If there is an active text editor look for Cargo.toml up to workspace's directory.
> If there is Cargo.toml use it.

Yes. Right now (in master branch, not in this PR) this is exactly how it works. This PR doesn't change it at all. That's why I'm saying that there is no performance hit in my approach.

> If there is no Cargo.toml use the last Cargo.toml.

Not sure about this (not trying WS root first), but I almost agree. 

If I open the file that belongs to the outside project (not in the WS) this project will be marked as last. After I close it, will extension still use the outside project instead of using the workspace project? This behaviour will be reverted to normal after I'll open any file from the workspace, but still...

> If there is no last Cargo.toml show an error.

I don't agree with this one. I think, that before showing an error we have to do the recursive search of Cargo.toml in the whole project.

Reason: this is a case where we can do a good guess that will work most of the time. It's better than the error, even if there will be one-time delay for a file search (async search, so it shouldn't be noticeable).

Otherwise my case still persists: close all tabs and you can't go to a symbol (in project w/o Cargo.toml in root). Yes, instead of an exception user will see an error message, but it's basically the same.

> If there is no active text editor use the last Cargo.toml.
>If there is no last Cargo.toml show an error.

If there is no active editor we do the same as if there is:
1. try workspace root
2. try last cwd
3. try recursive search

I don't see why the case, when there is no editor opened, should be any different from the case when the current file doesn't belong to a Rust project.

KalitaAlexey commented:
I haven't read all, but
> My case is about much more common example, when you are already working with the project and want to go to the symbol. Right now you can do it if you have a tab open. If there are no tabs open, extension fails.
This is a bug. It must be fixed.

mitskevich commented:
Yep, that's why I sent this PR ;) To fix this bug.

KalitaAlexey commented:
> I don't agree with this one. I think, that before showing an error we have to do the recursive search of Cargo.toml in the whole project.

Do you want it just for symbol search or also for commands?

Do you know Russian? Probably we can discuss it better in Russian.

mitskevich commented:
I want to do this for every project directory resolution (including commands). 

Let's keep public conversation in english, please contact me by email (in my profile), we can discuss privately in russian.

mitskevich commented:
Per our discussion, I have modified PR as follows:
* Using last cwd by default if it is not possible to determine cwd from opened editor. I decided to not save workspace directory as last cwd on project init, to keep everything path-related in one place. Workspace Cargo.toml check is just used as a fallback, it should have the same effect.
* Removed recursive search fallback and related code.
* Prevented remembering cwd if it is outside current workspace.

KalitaAlexey commented:
I will check it as soon as possible.

KalitaAlexey commented:
I have thought about our last resolution.
I see a one unresolved question, namely, which directory should the extension use in the following case:
I have the following structure:
```
Projects/
    Project 1/
      Cargo.toml
      src/
          main.rs

    Project 2/
        Cargo.toml
        src/
            main.rs
```
I open the **Projects** directory in an empty window of VSCode.
I open a new tab with the "Project 1/src/main.rs" file.
I build the project containing the file.
I open a Rust file located outside of the **Projects** directory.
I build the project containing the file.
I open a file located outside of the **Projects** directory.
I try to run a build command or a symbol search.

What will I get?

KalitaAlexey commented:
I am inclined to make a user open a project of a file instead of the file.
It restricts a little, but it makes the extension be intuitively.
I want you to contemplate about your motivation.
I open a Rust file.
I shouldn't be allowed to build a project containing the file.
I should be allowed to look for a symbol. We can make it (**rustsym** supports `LocalChildren`).
I should get an error if I would try to look for a symbol and **rustsym** isn't in PATH.

mitskevich commented:
> I open the Projects directory in an empty window of VSCode.
> I open a new tab with the "Project 1/src/main.rs" file.
> I build the project containing the file.
> I open a Rust file located outside of the Projects directory.
> I build the project containing the file.
> I open a file located outside of the Projects directory.
> I try to run a build command or a symbol search.
> What will I get?

If the last file is a Rust file belonging to some outside Rust project, you'll build that project or find all symbols in it. Is it incorrect behavior? That's what I would expect if I would be a user doing those actions.

> I open a Rust file.
> I shouldn't be allowed to build a project containing the file.

Why not?

> I should be allowed to look for a symbol. We can make it (rustsym supports LocalChildren).

I would like that too. It's another issue/PR though.

KalitaAlexey commented:
> If the last file is a Rust file belonging to some outside Rust project, you'll build that project or find all symbols in it. Is it incorrect behavior? That's what I would expect if I would be a user doing those actions.

No, it is a text file.

> Why not?

Because I opened a file, not a project. A user expects this behavior.

> I would like that too. It's another issue/PR though.

Yes, of course.

Our main problem is to decide what a user should be allowed to do when the one opens a Rust file.

I am not against symbol search, but I am against command execution.

mitskevich commented:
> No, it is a text file.

Then you will build or search symbols in "Project 1/". Again, I don't see any overly surprising behaviour here. I opened my project, then opened some outside lib file, played with it a bit, closed it, then opened some doc from my Downloads folder - I'm still working with my project.

> Because I opened a file, not a project. A user expects this behavior.

When I open Rust file, I expect linting (clippy/check) and symbol search. Doesn't matter if I opened one file or a folder.

Now for a build command - why disable it in this case? If one finds it convenient, let them use it. It's certainly will not be unexpected for them.

KalitaAlexey commented:
> Then you will build or search symbols in "Project 1/". Again, I don't see any overly surprising behaviour here. I opened my project, then opened some outside lib file, played with it a bit, closed it, then opened some doc from my Downloads folder - I'm still working with my project.

I didn't say I closed the Rust file located outside of the workspace's directory.

> When I open Rust file, I expect linting (clippy/check) and symbol search. Doesn't matter if I opened one file or a folder.

You shouldn't expect it. Please open a file in some IDE and check what it allows you.

I don't want to break accepted conventions.

mitskevich commented:
> I didn't say I closed the Rust file located outside of the workspace's directory.

Doesn't matter, you will build/search "Project 1/". It's expected behaviour, you are still working with your project.

> You shouldn't expect it. Please open a file in some IDE and check what it allows you.

VSCode is an editor, not IDE. It's closer to Sublime Text and Atom than Visual Studio. Most people who work with it expect more relaxed project-related behaviour. You should be able to do ninja stuff, quickly open unrelated files, experiment.

Look at the https://code.visualstudio.com/ or docs - there is no "IDE" combination anywhere (but a lot of "code editing").

KalitaAlexey commented:
> Doesn't matter, you will build/search "Project 1/". It's expected behaviour, you are still working with your project.

But before I executed command in context of the project located outside.
Doesn't it confuse?

> "code editing"

Exactly. Code editing.
I am up to racer support (autocompletion, hover) and rustsym support (symbol search) for a single file.
I am against a cargo command invocation.

mitskevich commented:
> But before I executed command in context of the project located outside.
> Doesn't it confuse?

If you have a project opened and drag in a file from another project, you would expect your project to be the context for everything except that file.

> I am against a cargo command invocation.

I still don't see why. You said it's because of conventions, but if there are any, they are certainly not what IDE-world is used to. I don't think this artificial restriction on building files from external projects has any usefulness or adds any clarity. Especially if it will affect linting (check/clippy are cargo commands too) - in that case it's downgrade.

KalitaAlexey commented:
@mitskevich,

Currently I have no time for this.
I will send you a message.

Bye

KalitaAlexey commented:
@mitskevich,
Will you have a time to prepare the PR?

mitskevich commented:
Per our discussion, removed attempt to check the workspace root directory.

KalitaAlexey commented:
@mitskevich,
I will check soon

KalitaAlexey commented:
Hi @mitskevich,
I don't like the chosen approach.
Also I see unrelated to the PR changes.
Do you mind if I implement the feature myself according to #36?

mitskevich commented:
Of course, no problem!
